pkg/backend/httpstate/token_source.go

```diff
@@ -0,0 +1,135 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package httpstate
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/logging"
+)
+
+// tokenSource is a helper type that manages the renewal of the lease token for a managed update.
+type tokenSource struct {
+	requests chan tokenRequest
+	done     chan bool
+}
+
+type tokenRequest chan<- tokenResponse
+
+type tokenResponse struct {
+	token string
+	err   error
+}
+
+func newTokenSource(
+	ctx context.Context,
+	initialToken string,
+	duration time.Duration,
+	refreshToken func(
+		ctx context.Context,
+		duration time.Duration,
+		currentToken string,
+	) (string, error),
+) (*tokenSource, error) {
+	requests, done := make(chan tokenRequest), make(chan bool)
+	ts := &tokenSource{requests: requests, done: done}
+	go ts.handleRequests(ctx, initialToken, duration, refreshToken)
+	return ts, nil
+}
+
+func (ts *tokenSource) handleRequests(
+	ctx context.Context,
+	initialToken string,
+	duration time.Duration,
+	refreshToken func(
+		ctx context.Context,
+		duration time.Duration,
+		currentToken string,
+	) (string, error),
+) {
+
+	renewTicker := time.NewTicker(duration / 8)
+	defer renewTicker.Stop()
+
+	state := struct {
+		token   string    // most recently renewed token
+		error   error     // non-nil indicates a terminal error state
+		expires time.Time // assumed expiry of the token
+	}{
+		token:   initialToken,
+		expires: time.Now().Add(duration),
+	}
+
+	renewUpdateLeaseIfStale := func() {
+		if state.error != nil {
+			return
+		}
+
+		now := time.Now()
+
+		// We will renew the lease after 50% of the duration
+		// has elapsed to allow time for retries.
+		stale := now.Add(duration / 2).After(state.expires)
+		if !stale {
+			return
+		}
+
+		newToken, err := refreshToken(ctx, duration, state.token)
+		// If renew failed, all further GetToken requests will return this error.
+		if err != nil {
+			logging.V(3).Infof("error renewing lease: %v", err)
+			state.error = fmt.Errorf("renewing lease: %w", err)
+			renewTicker.Stop()
+		} else {
+			state.token = newToken
+			state.expires = now.Add(duration)
+		}
+	}
+
+	for {
+		select {
+		case <-renewTicker.C:
+			renewUpdateLeaseIfStale()
+		case c, ok := <-ts.requests:
+			if !ok {
+				close(ts.done)
+				return
+			}
+			// If ticker has not kept up, block on
+			// renewing rather than risking returning a
+			// stale token.
+			renewUpdateLeaseIfStale()
```

Comment:
Oof - i've tried adding a test for this case, but it seems like it's not possible to reach here. The problem is that we need an active token to renew a lease. If we're detecting we've not kept up with refreshing tokens, we have a stale token, then we won't be able to review it either.  There is some grace here perhaps that the staleness check indicates a need to renew slightly before actual expiration. So if the ticker fails completely, this safeguard might be beneficial to keep. How to test this though..